rust-src/concordium_base/src/id/sigma_protocols/vcom_eq.rs

```diff
@@ -0,0 +1,413 @@
+//! The module provides the implementation of the `vcom_eq` sigma
+//! protocol. This protocol enables one to prove knowledge of
+//! $x_1, ..., x_n, r, (r_i)_{i\in I}$ such that $ C = h^r \prod_{i=1}^n
+//! g_i^{x_i}$ and $C_i = \bar{g}^{x_i} \bar{h}^{r_i}$ for $i\in I$.
+use super::common::*;
+use crate::{
+    common::*,
+    curve_arithmetic::{multiexp, Curve},
+    random_oracle::{Challenge, RandomOracle},
+};
+use ff::Field;
+use itertools::izip;
+use std::{collections::BTreeMap, rc::Rc};
+
+type IndexType = u8;
+
+/// The public information known to both the prover and the verfier.
+pub struct VecComEq<C: Curve> {
+    /// The commitment C
+    pub comm:  C,
+    /// The commitments C_i for i in I
+    pub comms: BTreeMap<IndexType, C>,
+    /// The points g_i references in the module description, in the given order.
+    pub gis:   Vec<C>,
+    /// The generator h
+    pub h:     C,
+    /// The generator \bar{g}
+    pub g_bar: C,
+    /// The generator \bar{h}
+    pub h_bar: C,
+}
+
+/// VComEq witness. We deliberately make it opaque.
+#[derive(Debug, Serialize)]
+pub struct Witness<C: Curve> {
+    #[size_length = 4]
+    sis: Vec<C::Scalar>,
+    t:   C::Scalar,
+    tis: BTreeMap<IndexType, C::Scalar>,
```

Comment:
Shouldn't we use `map_size_length` here as well?

Reply:
Yes, I think so. Will add.